### DISCUSSION TITLE
Why does txn.changes include resource that's not really changed?

### DISCUSSION THREAD

**Discussion by 0xbe1** [aptos-labs/aptos-developer-discussions#488] [Questions] [answered]
### Describe your question in detail.

txn1: https://explorer.aptoslabs.com/txn/1696780755/changes?network=mainnet
txn2: https://explorer.aptoslabs.com/txn/1696785786/changes?network=mainnet

both txns have `0x1::fungible_asset::Metadata` recorded as changed under `0xa` but it is clear that the Metadata struct never changes from txn1 to txn2.

what does a "change" really mean on Aptos?

---

originally posted at https://x.com/_0xbe1/status/1845140779989385428

### Which operating system are you using?

N/A

### Which SDK or tool are you using? (if any)

N/A

**Comment by ying-w** [comment, marked as answer]
Internally, resources are stored at the grain of 'State Key Hash'. As an optimization, we group together multiple resource types under the same state key as a resource group.

However, that means that if one type in the state key (resource group) changes, all types with the same state_key_hash is emitted. This has the convenience of also doing easier lookups

since fa::metadata is stored in same state key (resource group) as fa::supply so when supply changes as in the example you gave, metadata will also be emitted even though it doesn't change since it's all stored in the same state key

**Reply by 0xbe1** [reply]
@ying-w any chance we only emit the resource that's **REALLY** updated?

**Reply by ying-w** [reply]
Internally it's just a storage slot keyed on state_key_hash so the storage slot *is* being updated/rewritten

Yeah, it's messy, for resources, you need to get previous state and compare.

One approach would be to [get_account_resource](https://fullnode.mainnet.aptoslabs.com/v1/spec#/operations/get_account_resources) from API and parameterize `ledger_version` to current_version-1 to get previous state and then canonicalize the json and hash it to find differences

Another approach would be to go off [write_set_changes](https://github.com/aptos-labs/aptos-indexer-processors/blob/4cc90a3022b4736ce6f8c02fd55f65c13ed0bb1e/rust/processor/src/db/common/models/default_models/parquet_write_set_changes.rs), filter on resource changes (write_resource, delete_resource) and create table of state_key, resource_address, resource_type, resource_data (sorted), data_hashed to know when resources were last changed (and would also allow you to infer deletes on types since delete_resource only occurs when entire slot / state_key_hash is deleted). this is a kinda painful table to update

i've asked blockchain team to store previous_version for each of their states but gotten pushback since it would change schema (and only affects resources because of type, table items and modules is 1:1)

I think the approach we want to push devs towards is to have events rather than doing diffs on resources

**Reply by 0xbe1** [reply]
would resources live under the same slot because of the `ObjectGroup` annotation?

```
#[resource_group_member(group = aptos_framework::object::ObjectGroup)]
struct Metadata { ... }

#[resource_group_member(group = aptos_framework::object::ObjectGroup)]
struct ConcurrentSupply { ... }
```

**Reply by ying-w** [reply]
yes, everything under the same resource_group annotation will share the same state_key_hash
